### PR TITLE
fix: context plumbing & cancelation in client and server

### PR DESCRIPTION
**Pull request by guseggert**

### CODE REVIEW

File: blueprints/services/dagjson-over-http/server.go [hunk 1]
```diff
@@ -68,31 +67,42 @@ func (x GoServerImpl) GoDef() cg.Blueprint {
...
-				}
-				buf.WriteByte("\n"[0])
-				writer.Write(buf.Bytes())
-				if f, ok := writer.({{.HTTPFlusher}}); ok {
-					f.Flush()
+
+			writer.WriteHeader(200)
+			if f, ok := writer.({{.HTTPFlusher}}); ok {
+				f.Flush()
+			}
```
Comment (aschmahmann) [inline blueprints/services/dagjson-over-http/server.go:81]: Why are we writing out the header and flushing here? Trying to be nice to the client, or something having to do with making context cancellation easier?

Comment (guseggert) [reply blueprints/services/dagjson-over-http/server.go:81]: It's mostly about synchronization during tests, the client call doesn't return the channel to the caller until the HTTP response headers are received. Before this change, the server doesn't send the headers until the first result is written and flushed, which is surprising behavior (and confusing and annoying during tests), so explicitly writing the response code and flushing the headers causes the client to always return the channel before any results are even sent.

**Comment by aschmahmann** [comment]
looks reasonable to me, although can't review until later today. Perhaps @petar can

File: blueprints/services/dagjson-over-http/server.go (hunk 1 above)
Comment (petar) [reply blueprints/services/dagjson-over-http/server.go:81]: In a parallel PR https://github.com/ipld/edelweiss/pull/48, when a method is marked cachable it attaches an ETag to the response header. This cannot be done until after the request body is generated (below). It sounds like this optimization here won't work for that case?

Comment (guseggert) [reply blueprints/services/dagjson-over-http/server.go:81]: (I think you meant response body, not request body?) If we can't flush headers until the entire response body is generated, then we need to buffer the whole response and all of this becomes much easier. Is that what we want?

Comment (petar) [reply blueprints/services/dagjson-over-http/server.go:81]: Sounds good.

**Review by petar — APPROVED** [review]

**Pull request merged**